### PR TITLE
[Bandcamp] Extract additional fields

### PR DESCRIPTION
**Pull request by gkoelln**
* [bandcamp] Add additional extraction fields

Adding additional extraction fields

## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Adding additional extraction fields for tracks, album, and artist information

### CODE REVIEW

File: youtube_dl/extractor/bandcamp.py [hunk 1]
```diff
@@ -233,5 +256,6 @@ def _real_extract(self, url):
             'uploader_id': uploader_id,
             'id': playlist_id,
             'title': title,
+            'album': title,
             'entries': entries,
         }
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: All these fields must be optional.

Comment (gkoelln) [inline youtube_dl/extractor/bandcamp.py]: I'm afraid I don't know what that means.

Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: That's explained in code conventions you've ticked as read.

Comment (gkoelln) [inline youtube_dl/extractor/bandcamp.py]: Fair enough.  So will adding fatal=False be enough here?

Comment (gkoelln) [inline youtube_dl/extractor/bandcamp.py]: Added more edits.  Am I getting closer?

Comment (gkoelln) [inline youtube_dl/extractor/bandcamp.py]: One more edit since the last comment. It all tested well, from what I can tell.

Comment (gkoelln) [inline youtube_dl/extractor/bandcamp.py]: Added track_id to free download

File: youtube_dl/extractor/bandcamp.py [hunk 2]
```diff
@@ -77,6 +86,12 @@ def _real_extract(self, url):
                     'title': data['title'],
                     'formats': formats,
                     'duration': float_or_none(data.get('duration')),
+                    'track': data.get('title'),
+                    'track_number': data.get('track_num'),
+                    'track_id': track_id,
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: It's not a track id.

File: youtube_dl/extractor/bandcamp.py [hunk 3]
```diff
@@ -77,6 +86,12 @@ def _real_extract(self, url):
...
+                    'album': album,
+                    'album_artist': album_artist,
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: It's artist not an album artist.

File: youtube_dl/extractor/bandcamp.py [hunk 4]
```diff
@@ -86,6 +101,10 @@ def _real_extract(self, url):
             r'(?ms)var TralbumData = .*?[{,]\s*id: (?P<id>\d+),?$',
             webpage, 'video id')
 
+        track_number = self._search_regex(
+            r'"track_num":(?P<track_number>\d+),',
+            webpage, 'track number', fatal=False)
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: Must be int.

File: youtube_dl/extractor/bandcamp.py [hunk 5]
```diff
@@ -77,6 +86,12 @@ def _real_extract(self, url):
                     'title': data['title'],
                     'formats': formats,
                     'duration': float_or_none(data.get('duration')),
+                    'track': data.get('title'),
+                    'track_number': data.get('track_num'),
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: Must be int.

Comment (gkoelln) [inline youtube_dl/extractor/bandcamp.py]: Understood.

File: youtube_dl/extractor/bandcamp.py (hunk 2 above)
Comment (gkoelln) [inline youtube_dl/extractor/bandcamp.py]: Then what's the difference between id and track_id?

File: youtube_dl/extractor/bandcamp.py (hunk 3 above)
Comment (gkoelln) [inline youtube_dl/extractor/bandcamp.py]: "album_artist: List of all artists appeared on the album"
Based on that definition, I guess you're right.  I was thinking album artist vs. track artist (like on a mixtape, or similar).

File: youtube_dl/extractor/bandcamp.py (hunk 4 above)
Comment (gkoelln) [inline youtube_dl/extractor/bandcamp.py]: Understood.

**Comment by gkoelln** [comment]
Changes made.

**Comment by dstftw** [comment]
Squash commits.

**Comment by gkoelln** [comment]
Sorry. Still learning.

**Comment by gkoelln** [comment]
Commits have been squashed, by the way.

File: youtube_dl/extractor/bandcamp.py [hunk 6]
```diff
@@ -47,6 +47,15 @@ def _real_extract(self, url):
         mobj = re.match(self._VALID_URL, url)
         title = mobj.group('title')
         webpage = self._download_webpage(url, title)
+        album = self._search_regex(
+            r'(?ms).*?title\s*?:\s*?"(?P<album>.*?)",',
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: This extracts first title found not necessarily an album title.
All captured groups must be at least one character length.

File: youtube_dl/extractor/bandcamp.py [hunk 7]
```diff
@@ -47,6 +47,15 @@ def _real_extract(self, url):
...
+            webpage, 'album', fatal=False)
+        artist = self._search_regex(
+            r'(?ms)var EmbedData = .*?[{,]\s*artist:\s*?"(?P<artist>.*?)",$',
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: What's the point of this noise `var EmbedData = .*?[{,]\s*`?

File: youtube_dl/extractor/bandcamp.py [hunk 8]
```diff
@@ -47,6 +47,15 @@ def _real_extract(self, url):
...
+            webpage, 'artist', fatal=False)
+        release_year = self._search_regex(
+            r'(?ms).*?release_date"?:\s*?"\d+ \w+ (?P<release_year>\d+)\s*?.*?GMT",',
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: `utils.unified_strdate`, `release_date`.

File: youtube_dl/extractor/bandcamp.py (hunk 6 above)
Comment (gkoelln) [inline youtube_dl/extractor/bandcamp.py]: So, something like this?
```
album = self._search_regex(
    r'(?ms).*?album_title\s*?:\s*?"(?P<album>[^"]+)",',
```
Also, if a track is just a single track and not part of an album (such as the youtube-dl test song), do we just want to leave album as NA or have the album name be the same as the track title?  (Perhaps that's just a personal preference of some?)

File: youtube_dl/extractor/bandcamp.py (hunk 7 above)
Comment (gkoelln) [inline youtube_dl/extractor/bandcamp.py]: I wanted to use the artist that was included in the EmbedData var (which I can now see is the first instance of artist, usually). In most cases, three of four instances of the artist are the same, with one being null (in TralbumData).  Should I just remove the artist field altogether?

File: youtube_dl/extractor/bandcamp.py (hunk 6 above)
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: It should extract the album title field reported by the webpage. No more no less.

File: youtube_dl/extractor/bandcamp.py (hunk 7 above)
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: If they are all the same extract any that is not null.

**Comment by gkoelln** [comment]
Changes made.

**Pull request closed without merging**